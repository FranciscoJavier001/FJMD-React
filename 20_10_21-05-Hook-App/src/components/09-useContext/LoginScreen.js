//** Va a tener un login, solo quiero que aqui se grabe lo que es el usuario */

import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const LoginScreen = () => {

    const { setUser } = useContext( UserContext );

    return (
        <div>
            <h1>LoginScreen</h1>
            <hr />
            <button 
                className="btn btn-primary"
                onClick={ ()=> setUser({
                    id: 123,
                    name: 'Fernando'
                })}
            >
                Login
            </button>
        </div>
    )
}

