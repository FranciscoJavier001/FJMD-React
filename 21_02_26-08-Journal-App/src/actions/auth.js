//** Estas son acciones que vamos a definir */
import { firebase, googleAuthProvider } from '../firebase/firebase-config'
import { types } from '../types/types'

//** Esto es una tarea asincrona, osea que se va a ejecutar todo el proceso, luego se va a encontrar con el dispatch que ejecuta la accion del login que es la que se va a aplicar directamente en el store (la fuente unica de la verdad), que modifica el state y nos regresa todo lo que estamos esperando  */
export const startLoginEmailPassword = (email, password) => { //** Esta funcion va a recibir el email y el password */
    return (dispatch) => { //** Esta funcion va a regresar un callback, osea una funcion a disparar */
        //** Hay que llamar a otro dispatch cuando ya se ejecute todo, cuando ya tenga la data */
        setTimeout(() => {
            dispatch( login( 123, 'Pedro')) //** Aqui voy a hacer un dispatch del login, con el usuario */
        }, 3500);
    }
}

export const startRegisterWithEmailPasswordName = ( email, password, name ) => { //** Esta es una tarea asincrona, asi que voy a tener un callback */
    return ( dispatch ) => { //** Cuando tenga el usuario grabado en firebase voy a hacer el dispatch */
        firebase.auth().createUserWithEmailAndPassword( email, password ) //** Aqui mandamos los parametros de email y password */
        .then( async ({ user }) => { //** Esta va a ser una promesa */

            await user.updateProfile({ displayName: name }) //** que me de el name que estoy recibiendo como argumento de mi funcion y esto me regresa una promesa */

            dispatch( //** Esto es para llamar de nuevo el dispatch */
                login( user.uid, user.displayName)
            )
        })
        .catch( e => { //** Esto es para atrapar el error y mostrarlo */
            console.log(e);
        })
    }
}

export const startGoogleLogin = () => { //** Una funcion de fleca que no recibe nada, pero es una tarea sincrona voy a poner el return y el callback me va a proveer el dispatch, por lo cual cuando tenga la informacion lo voy a llamar */
    return ( dispatch ) => {
    firebase.auth().signInWithPopup( googleAuthProvider ) //** Esto va a retormarme una promesa */
        .then( ({ user }) => { //** Aqui voy a usar la desestructuracion para extraer el user */
            dispatch( //** Que cuando esto se responda voy a hacer el dispatch de la accion que ya tengo definido abajo con el login, donde voy a poder el id que viene del user.uid, y el displayName que viene del user.displayName */
                login( user.uid, user.displayName)
            )
        })
    }
}

export const login = (uid, displayName) => ({ //** Esta es la accion que voy a llamar cuando tenga el uid y el displayNameny esta accion me debe regresar un type.login */
    type: types.login,
    payload: { //** El payload va a ser un objeto que tenga un uid y el displayName */
        uid,
        displayName
    }
})