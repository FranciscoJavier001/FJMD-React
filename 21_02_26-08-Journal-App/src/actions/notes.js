import { db } from "../firebase/firebase-config";
import { types } from '../types/types'

//** Voy a exportar un nuevo componente */
export const startNewNote = () => { //** Esta es una tarea asyncrona, asi que voy a poner el return con un callback que va a ocupar el dispatch, y lo voy a llamar en el dispatch del Sidebar.js en el handleAddNew */
    return async( dispatch, getState ) => { //** Aqui voy a poner el segundo argumento, despues del dispatch que vamos a llamar getState (pero ese segundo argumento es una funcion para obtener el state similar al useSelector) */
        //** Para grabar en Firestore voy a oupar otras cosas como el ID del usuario */
        const uid = getState().auth.uid //** Este es todo el State */
        console.log( uid );

        //** Ahora voy a crear la nueva nota que quiero guardar */
        const newNote = {
            title: '', //** Este lo voy a inicializar vacio */
            body: '',
            date: new Date().getTime() //** Este me va a dar el tiempo exacto cuando la nota fue creada, es un objeto de JS, y lo voy a mandar a un path de Firestore para guardarlo */
        }

        //** Para hacer lo del getTime, debo de conseguir la referencia a la base de datos, esta ya la tengo en firebase & firebase-config.js */
        const doc = await db.collection(`${ uid }/journal/notes`).add( newNote ) //** Voy a hacer la referencia al documento y esto viene de db que lo explique arriba, esto lo voy a poner en la coleccion donde quiero guardar esto, voy a ocupar el uid del usuario , luego el journal que va a ser donde lo estoy guardando, luego notes, porque quiero que ahi este y ahi ya puedo hacer el .add y ahi mando el newNote, el add regresa una promesa que retorna el document.reference y como esto es una promesa en el return del startNewNote lo voy a convertir a async y aqui hacer el await, osea espera que se haga la insercion y el doc va a dar referencia al documento */

        //** Aqui voy a mandar llamar la accion que defini abajo */
        dispatch( activeNote( doc.id, newNote ) ) //** Esta pide el id de la nota y este lo tengo en el doc.id, y luego viene la nota que seria el newNote que lo acabo de crear */
    }
}

//** Lo ultimo despues de todo lo que hemos echo es manejar esta accion el noteReducer */
//** Voy a crear una nueva accion que sea syncrona porque ya voy a tener la informacion voy a mandar algo al reducer  */
export const activeNote = ( id, note ) => ({ //** Este va a recibir el id y la nota */
    //** Este objeto va a tener los types */
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
})

//** Como voy a guardar las notas en el store voy a guardar esto, con esto estoy creando siempre un nuevo arreglo */
export const setNotes = ( notes ) => ({ //** Aqui voy a recibir las notas */
    type: types.notesLoad, //** el type que va a tener esto , va a ser de notesLoad y el paylod va a ser todas las notas que este recibiendo como argumento */
    payload: [ notes ]
})