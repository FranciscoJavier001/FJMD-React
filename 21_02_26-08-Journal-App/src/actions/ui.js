//** Acciones en el UI */

import { types } from "../types/types";

export const setError = ( err ) => ({ //** De una vez voy a hacer el return del objeto y va a recibir el payload del error */
    type: types.uiSetError,
    payload: err
})

export const removeError = () => ({ //** De una vez voy a hacer el return del objeto y no va a recibir ningun objeto */
    type: types.uiRemoveError, //** Va a llamar a la funcion de remover error */
})

export const startLoading = () => ({ //** Asi llamamos las acciones, y ahora se debe hacer el dispatch de estas acciones */
    type: types.uiStartLoading,
})

export const finishLoading = () => ({
    type: types.uiFinishLoading,
})