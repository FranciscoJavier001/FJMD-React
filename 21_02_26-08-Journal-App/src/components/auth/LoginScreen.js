import React from 'react'
import { Link } from 'react-router-dom'

export const LoginScreen = () => {
    return (
        <>
            <h3 className="auth__title">Login</h3>  {/* Es lo que esta escrito en la caja blanca definida en el auth__box-container que contiene a los 2 */}

            <form> {/* Este es la caja de un formulario que va a estar abajo del login */}
                <input //** Este crea el primero */
                    type="text"
                    placeholder="E-Mail"
                    name="email"
                    className="auth__input"
                    autoComplete="off" //** Para que no me este dando sugerencias ni nada al escribir */
                />

                <input //** Este crea el segundo */
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                />

                <button
                    type="submit" //** Que pueda darle click */
                    className="btn btn-primary btn-block"
                >
                    Login
                </button>

                <div className="auth__social-network">
                    <p>Login with Social Networks</p>
                    <div 
                        className="google-btn"
                        >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Login with Google</b>
                        </p>
                    </div>
                </div>

                <div className="link-style"> {/* Asi creamos una clase para implementarla en el SCSS */}
                <Link to="/auth/register" /* Este lo importamos desde react-router-dom con el link */
                >
                    Create New Account
                </Link>
                </div>

            </form>
        </>
    )
}