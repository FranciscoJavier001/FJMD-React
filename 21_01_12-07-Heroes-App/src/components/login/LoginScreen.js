//**_______________________________________________________________________________________________________________________________________________*/
import React, { useContext } from 'react'
import { AuthContext } from '../../auth/AuthContext';
import { types } from '../../types/types';

export const LoginScreen = ({ history }) => { //** El LS recibe el history(este vide de Navegador/Components/props) */

    
    const { dispatch } = useContext( AuthContext ); //** Tambien esta en src/HeroesApp y en el dispatch vienen parametros que mandamos/recibimos */

    const handleLogin = () => { //** Funcion que no recibe nada */
        // history.push('/'); //** Ruta a la que voy a navegar con click en el login */
        // history.replace('/'); //** Asi no hay history, y solo me lleva directamente al "/" y no puedo regresar al login manualmente */

        //** La ultimaRuta, la sacamos del lS, que nos manda a la ultima pagina visitada o si no al "/", autentificado a marvel y si no al login*/
        const lastPath = localStorage.getItem('lastPath') || '/'; 
        
        dispatch({ //** Disparo la accion, que defini en los types, y ahi tengo contexto, paso estas variables aqui definidas con auth */
            type: types.login, //** Al hacer el login va a poner el name como Francisco con los types, que lo defini l.82 components/ui/Navbar */
            payload: {
                name: 'Francisco'
            }
        });

        history.replace( lastPath ); //** Al entrar se remplaza todo, para ya no poder acceder de nuevo al login, hasta hacer el logout */
    }

    return (
        <div className="container mt-5"> {/* Container, lo pone medio centrado y el mt-5 es el espacio que tiene hacia arriba */}
            <h1>Login</h1>
            <hr/>

            <button
                className="btn btn-primary" //** btn=Boton, b-p=Color azul */
                onClick={ handleLogin } //** Metodo que se dispara*/
            >
                Login
            </button>
        </div>
    )
}
