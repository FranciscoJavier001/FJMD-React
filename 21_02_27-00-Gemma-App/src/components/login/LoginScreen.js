import React, { useContext } from 'react'
import { AuthContext } from '../../auth/AuthContext';
import { types } from '../../types/types';

export const LoginScreen = ({ history }) => { //** Vamos a usar la desestructuracion y ahi vamos a tener el history */

    //** 1. Para la tarea lo primero que debo hacer es extraer el context, de aqui lo que me interesa es extraer el dispatch y en el useContext el AuthContext */
    const { dispatch } = useContext( AuthContext );

    const handleLogin = () => { //** Esta es una funcion de flecha, que no recibe nada */
        // history.push('/'); //** Asi le ponemos que va a hacer cuando toque el boton */
        // history.replace('/'); //** Investigar */

        const lastPath = localStorage.getItem('lastPath') || '/'; //** Voy a crear una variable llamada lastPath que va a ser el lastPath que pusimos en la Ruta Privada y por si paso cualquier cosa es la primera vez que entra, porque purgo o algo asi que lo redireccione al replace */

        //** Tarea.. Del ontext extraer el dispatch y ese dispatch tiene que mandar un action que tenga el tipo del case types.login y case types logout, el objeto que vamos a mandar en el payload de la accion va a ser algo sencillo como name: 'Francisco', es todo lo que necesito en el dispatch, esto deberia de establecerselo al reducer para ver en components en los hooks el logged en true y lo que colocamos en el name, y se va a mantener porque vamos a tomar todo lo que venga en el payload de la accion y tambien la propiedad logged que ya se autentifico */

        //** 2. Ahora voy a disparar esa accion, voy a ser el dispatch de un objeto que tiene un type y el types viene de types.login y luego viene el payload: { (y va a ser un objeto que tiene el name y mi nombre)}, y abajo pongo el replace de la ruta (pero ya tengo mi login en un nivel alto de la app) */
        
        //** Cambiamos el lugar donde estaba el replace */
        // history.replace('/'); //** Este me hace que cuando le pongo el login, en automatico me meta a la aplicacion */
        dispatch({
            type: types.login,
            payload: {
                name: 'Francisco'
            }
        });

        history.replace( lastPath ); //** Este me hace que cuando le pongo el login, en automatico me meta a la aplicacion, esto se modifico en otro video, en el cual va a tener siempre guardada la ultima ruta que se visito */
    }

    return (
        // Container es un contenedor, que se pone donde le digamos, si le aignamos un margen de arriba, osea es una linea que se queda solo en lo que vamos a poner
        <div className="container mt-5">
            <h1>Login</h1>
            <hr/>

            <button
                className="btn btn-primary"
                onClick={ handleLogin } //** Este metodo hay que declararlo arriba, para que no nos de ninguna falla */
            >
                Login
            </button>
        </div>
    )
}
