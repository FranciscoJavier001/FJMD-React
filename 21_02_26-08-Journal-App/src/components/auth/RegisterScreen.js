import React from 'react'
import { Link } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import validator from 'validator'

import { useForm } from '../../hooks/useForm'
import { removeError, setError } from '../../actions/ui'
import { startRegisterWithEmailPasswordName } from '../../actions/auth'

export const RegisterScreen = () => {
    
    const dispatch = useDispatch() //** Usamos el dispatch para disparar la accion que queramos */
    const { msgError } = useSelector( state => state.ui) //** Este es un Hook y recibe un callback, en el que tengo el state y me va a regresar el state, y lo desestructire para solamente ver el mensaje de error */


    const [ formValues, handleInputChange ] = useForm({ //** Aqui usamos el useForm para el formulario */
        name: 'Andrea', //** Voy a mandar un objeto, para hacer el registro de usuario */
        email: 'andrea@email.com',
        password: '123456',
        password2: '123456',
    })

    //** El formValues, son los valores que tenemos en el formulario, y por eso de ahi se extraen y el handleInputChange es lo que va a quedar cuando se haga el cambio */
    const { name, email, password, password2 } = formValues //** ASi voy a tener la desestructuracion de objetos del formValues */

    const handleRegister = (e) => {
        e.preventDefault()
        
        if( isFormValid() ) { //** Hago el llamado a la funcion, que solo se regresa si es true */
            dispatch( startRegisterWithEmailPasswordName(email, password, name) ) //** Vamos a hacer el dispatch de la accion startRegisterWithEmailPasswordName, recuerda tenemos que importarlo y vamos a mandar todo lo que definimos  */
        }
    }

    const isFormValid = () => {
        if ( name.trim().length === 0 ) { //** Esto es si la persona tiene un nombre vacio que salga el mensaje de error*/
            dispatch( setError('Nombre es Requerido') ) //** Lo importamos de los typos */
            return false;
        } else if ( !validator.isEmail( email ) ) { //** Si esto no es un email, quiero que hagas esto */
            dispatch( setError('Email no valido') )
            return false;
        } else if ( password !== password2 || password.length < 5 ) {//** Si los password son diferentes o el password tiene menos de 6 digitos */
            dispatch( setError('Password por lo menos de 6 digitos y tiene que coincidir ambos') )
            return false;
        }
        dispatch( removeError() ) //** El caso opuesto de cuando hay error */
            return true;
    }

    return (
        <>
        <h3 className="auth__title">Register</h3>  {/* Es lo que esta escrito en la caja blanca definida en el auth__box-container que contiene a los 2 */}

        <form 
            onSubmit={ handleRegister }
            className="animate__animated animate__fadeIn animate__faster"
        > {/* Este es la caja de un formulario que va a estar abajo del login */}

            {
                msgError && //** Que solo se muestre si es diferente a null */
                ( //** Esto es el bloque de la condicion */
                    <div className="auth__alert-error"> {/* Que esto salga cuando hay un error en el formulario */}
                        { msgError }
                    </div>
                )
            }

            <input //** Este crea otro donde la gente puede escribir */
                type="text"
                placeholder="Name"
                name="name"
                className="auth__input"
                autoComplete="off" //** Para que no me este dando sugerencias ni nada al escribir */
                value={ name }
                onChange={ handleInputChange }
            />

            <input //** Este crea el primero */
                type="text"
                placeholder="E-Mail"
                name="email"
                className="auth__input"
                autoComplete="off" //** Para que no me este dando sugerencias ni nada al escribir */
                value={ email }
                onChange={ handleInputChange }
            />

            <input //** Este crea el segundo */
                type="password"
                placeholder="Password"
                name="password"
                className="auth__input"
                value={ password }
                onChange={ handleInputChange }
            />

            <input //** Este crea el segundo */
                type="password"
                placeholder="Confirm Password"
                name="password2"
                className="auth__input"
                value={ password2 }
                onChange={ handleInputChange }
            />

            <button
                type="submit" //** Que pueda darle click */
                className="btn btn-primary btn-block mb-5"
            >
                Register
            </button>

            <Link
                to="/auth/login" /* Este lo importamos desde react-router-dom con el link */
                className="link"
            >
                Already Register?
            </Link>

        </form>
    </>
    )
}