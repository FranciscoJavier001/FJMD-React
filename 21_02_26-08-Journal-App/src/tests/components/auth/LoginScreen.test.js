import React from 'react' //** Por que voy a utilizar JSX */
import { mount } from 'enzyme' //** Recuerda para usar el mount */
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom' //** Este me va a permitir definir las rutas, como estuviera navegando en las rutas, pero aqui adento */

import configureStore from 'redux-mock-store' //ES6 modules
import thunk from 'redux-thunk'

import '@testing-library/jest-dom'

import { LoginScreen } from "../../../components/auth/LoginScreen"
import { startGoogleLogin } from '../../../actions/auth'

//** Voy a aplicar un mock a la funcion para saber si se llamo o no, y en esta funcion de flecha voy a regresar un objeto y dentro voy a tener el startGoogleLogin */
jest.mock('../../../actions/auth', () => ({
    startGoogleLogin: jest.fn() //** Que esto no es mas que un jest.fn() */
}))


/** Crea los middleware, configura el mockStore */
const middlewares = [thunk] //** thunk es el middleware */
const mockStore = configureStore(middlewares)

const initState = { //** Este contiene el estado del store en este momento, pero lo vamos a desestructurar */
    auth: {}, //** Lo voy a tener como un objeto vacio */
    ui: { //** Y este es como lo voy a tener en el estado inicial del test */
        loading: false,
        msgError: null
    }
}

//** Voy a crear un store que es un objeto */
let store = mockStore(initState)
store.dispatch = jest.fn //** Asi vamos a evaluar como son llamadas las funciones, osea vamos a reemplazar la funcion que tenga el store por una funcion jest la cual si voy a poder evaluar y voy a tener control absoluto sobre la funcion */

const wrapper = mount( //** Voy a ocupar renderizar mas cosas */
    <Provider store={ store }> {/* Aqui voy a proveer el store que va a ser igual sl store del mockStore que se creo en la arriba en la linea 18 */}
        <MemoryRouter>
            <LoginScreen /> 
        </MemoryRouter>
    </Provider>
)

describe('Pruebas en LoginScreen />', () => {

    //** Esto es para hacer una limpieza, es decir lo va a reinicializar por si ocupo hacer mas pruebas en el store */
    beforeEach(()=>{
        store = mockStore(initState) //** El store va a ser igual al createMockStore */
        jest.clearAllMocks() //** Es buena practica limpiar el mook, para reestablecer cada uno de los valores */
    })
    
    test('debe de mostrarse correctamente', () => {
        
        expect( wrapper ).toMatchSnapshot(); 
    })

    test('debe de disparar la accion de startGoogleLogin', () => {
        
        //** Asi es para simular un click, y su propiedad es handleGoogleLogin, para eso uso el prop, y puedo hacer ua eferencia a la funcion */
        wrapper.find('.google-btn').prop('onClick')()

        //** Yo esperaria que esta funcion haya sido llamada */
        expect( startGoogleLogin ).toHaveBeenCalled()
    })
    
    
})