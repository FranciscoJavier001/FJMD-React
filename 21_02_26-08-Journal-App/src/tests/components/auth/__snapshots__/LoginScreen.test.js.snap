// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en LoginScreen /> debe de mostrarse correctamente 1`] = `
Array [
  " ",
  Array [
    <h3
      className="auth__title"
    >
      Login
    </h3>,
    "  ",
    <form
      className="animate__animated animate__fadeIn animate__faster"
      onSubmit={[Function]}
    >
       
      <input
        autoComplete="off"
        className="auth__input"
        name="email"
        onChange={[Function]}
        placeholder="E-Mail"
        type="text"
        value="frank@email.com"
      />
      <input
        className="auth__input"
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value="123456789"
      />
      <button
        className="btn btn-primary btn-block"
        disabled={false}
        type="submit"
      >
        Login
      </button>
      <div
        className="auth__social-network"
      >
        <p>
          Login with Social Networks
        </p>
        <div
          className="google-btn"
          onClick={[Function]}
        >
          <div
            className="google-icon-wrapper"
          >
            <img
              alt="google button"
              className="google-icon"
              src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
            />
          </div>
          <p
            className="btn-text"
          >
            <b>
              Login with Google
            </b>
          </p>
        </div>
      </div>
      <a
        className="link"
        href="/auth/register"
        onClick={[Function]}
      >
        Create New Account
      </a>
    </form>,
  ],
]
`;
